### PR TITLE
feat: Add OTel Gen AI semantic convention input / output messages

### PR DESCRIPTION
**Pull request by brightsparc**
Adds OpenTelemetry Gen AI Semantic Convention message attributes including:

`gen_ai.system_instructions`
`gen_ai.input.messages`
`gen_ai.output.messages`

Converts message formats (text, tool_use, tool_result, images) to the standardized semconv message part format.

### CODE REVIEW

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 1]
```diff
@@ -296,28 +571,48 @@ def on_response(response: ResponseT, span: LogfireSpan) -> ResponseT:
...
-            span.set_attribute('events', events)
+        response_output_messages: OutputMessages = convert_responses_outputs_to_semconv(response)
+        span.set_attribute(OUTPUT_MESSAGES, response_output_messages)
+        # Keep 'events' for backward compatibility
+        existing_events: list[Any] = []
+        otel_span = span._span  # pyright: ignore[reportPrivateUsage]
+        if otel_span is not None and hasattr(otel_span, 'attributes') and otel_span.attributes:
+            events_attr = otel_span.attributes.get('events')
+            if isinstance(events_attr, list):  # pragma: no cover
+                existing_events = cast(list[Any], events_attr)
```
[the remark this answers is not in the document]
Comment (brightsparc) [reply logfire/_internal/integrations/llm_providers/openai.py]: Made this changed and resolved merge conflicts.

**Comment by brightsparc** [comment]
@alexmojaki ready for final review from you.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 2]
```diff
@@ -184,6 +214,167 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
+    if part_type == 'text':
+        return TextPart(type='text', content=part.get('text', ''))
+    elif part_type == 'image_url':
+        url = part.get('image_url', {}).get('url', '')
+        return UriPart(type='uri', uri=url, modality='image')
+    elif part_type in ('input_audio', 'audio'):  # pragma: no cover
+        return BlobPart(
+            type='blob',
+            content=part.get('data', ''),
+            modality='audio',
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/_internal/integrations/llm_providers/openai.py:321]: I guess we don't have tests for audio, not even sure what it does right now. If tests are added, they should use VCR, not mocking.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 3]
```diff
@@ -194,6 +389,89 @@ def is_current_agent_span(*span_names: str):
...
+                    with contextlib.suppress(json.JSONDecodeError):
+                        func_args = json.loads(func_args)
+                parts.append(
+                    ToolCallPart(
+                        type='tool_call',
+                        id=tc.id,
+                        name=tc.function.name,
+                        arguments=func_args,
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (brightsparc) [reply logfire/_internal/integrations/llm_providers/openai.py:432]: The Devin review is incorrect. `ChatCompletionMessageToolCall` and `ChatCompletionMessageFunctionToolCall` are the **same class** — the SDK aliases them:

```python
>>> from openai.types.chat.chat_completion_message_function_tool_call import ChatCompletionMessageFunctionToolCall
>>> from openai.types.chat import ChatCompletionMessageToolCall
>>> ChatCompletionMessageFunctionToolCall is ChatCompletionMessageToolCall
True
```

The `tool_calls` field on `ChatCompletionMessage` is typed as `list[ChatCompletionMessageFunctionToolCall | ChatCompletionMessageCustomToolCall] | None`, so the `isinstance` check exists to filter out the `CustomToolCall` variant, which is intentional.

[the remark this answers is not in the document]
Comment (brightsparc) [reply logfire/_internal/integrations/llm_providers/openai.py:432]: @alexmojaki any more work to do this can we please get this landed?

File: logfire/_internal/integrations/llm_providers/anthropic.py [hunk 4]
```diff
@@ -97,6 +120,132 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
+        return TextPart(type='text', content=part.get('text', ''))
+    elif part_type == 'image':  # pragma: no cover
+        source = part.get('source', {})
+        if source.get('type') == 'base64':
+            return BlobPart(
+                type='blob',
+                modality='image',
+                content=source.get('data', ''),
+                media_type=source.get('media_type'),
+            )
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/anthropic.py]: Fixed in b20e915e — now only including `media_type` in the BlobPart when the value is actually present, respecting the `NotRequired[str]` type contract.

File: logfire/_internal/integrations/llm_providers/openai.py
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:1]: This is a known limitation — streaming chat completions don't accumulate output messages in this PR. The non-streaming path and the Responses streaming path both set OUTPUT_MESSAGES. Adding streaming chat completion output messages is left for a follow-up.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 5]
```diff
@@ -194,6 +389,89 @@ def is_current_agent_span(*span_names: str):
...
+                func_args: Any = tc.function.arguments
+                if isinstance(func_args, str):
+                    with contextlib.suppress(json.JSONDecodeError):
+                        func_args = json.loads(func_args)
+                parts.append(
+                    ToolCallPart(
+                        type='tool_call',
+                        id=tc.id,
+                        name=tc.function.name,
+                        arguments=func_args,
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:430]: Acknowledged — the isinstance check is intentional to filter out `ChatCompletionMessageCustomToolCall`. As brightsparc explained in the other thread, `ChatCompletionMessageFunctionToolCall` is the same class as `ChatCompletionMessageToolCall`.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 6]
```diff
@@ -184,6 +214,171 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
+    """Convert a single content part to semconv format."""
+    if isinstance(part, str):  # pragma: no cover
+        return TextPart(type='text', content=part)
+
+    part_type = part.get('type', 'unknown')
+    if part_type == 'text':
+        return TextPart(type='text', content=part.get('text', ''))
+    elif part_type == 'image_url':  # pragma: no cover
+        url = part.get('image_url', {}).get('url', '')
+        return UriPart(type='uri', uri=url, modality='image')
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:316]: Acknowledged — this is consistent with the existing `request_data` behavior. Splitting data URIs into BlobPart could be a future improvement but is out of scope for this PR.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 7]
```diff
@@ -184,6 +214,171 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
+                            'ChatMessage',
+                            {
+                                'role': 'assistant',
+                                'parts': [
+                                    ToolCallPart(
+                                        type='tool_call',
+                                        id=inp.get('call_id', ''),
+                                        name=inp.get('name', ''),
+                                        arguments=arguments,
+                                    )
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:375]: This has already been addressed — the current code at lines 344-348 in `convert_responses_inputs_to_semconv` and lines 452-455 in `convert_responses_outputs_to_semconv` both include JSON parsing for arguments.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 8]
```diff
@@ -184,6 +214,171 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
+                        with contextlib.suppress(json.JSONDecodeError):
+                            arguments = json.loads(arguments)
+                    parts.append(
+                        ToolCallPart(
+                            type='tool_call',
+                            id=tc.get('id', ''),
+                            name=function.get('name', ''),
+                            arguments=arguments,
+                        )
+                    )
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:290]: Acknowledged — these pragma: no cover branches handle edge cases (multimodal content, tool calls in input messages) that would need VCR-based tests with real API calls. Adding test coverage for these paths can be done in a follow-up.

**Comment by aditya-ai-architect** [comment]
Acknowledged, I will review this.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 9]
```diff
@@ -226,9 +530,14 @@ def get_response_data(self) -> Any:
...
+        versions = self._versions
         response = self.get_response_data()
         if response:
-            span_data['events'] = span_data['events'] + responses_output_events(response)
+            if 'latest' in versions:
+                output_messages = convert_responses_outputs_to_semconv(response)
+                span_data[OUTPUT_MESSAGES] = output_messages
+            if 1 in versions:
+                span_data['events'] = span_data.get('events', []) + responses_output_events(response)
         return span_data
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py:550]: Fixed in 7903fe3a — `OpenaiResponsesStreamState.get_attributes` now uses a shallow copy (`result = dict(**span_data)`) consistent with all other `get_attributes` implementations.

File: tests/otel_integrations/test_anthropic.py [hunk 10]
```diff
@@ -948,7 +948,6 @@ def test_request_parameters(instrumented_client: anthropic.Anthropic, exporter:
                             'cache_creation': None,
                             'cache_creation_input_tokens': None,
                             'cache_read_input_tokens': None,
-                            'inference_geo': None,
                             'input_tokens': 2,
                             'output_tokens': 3,
                             'server_tool_use': None,
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply tests/otel_integrations/test_anthropic.py:1048]: Fixed in 0a400544 — the `inference_geo` field has been added back to the snapshot. It was missing because the Anthropic SDK added this field to the Usage model.

File: logfire/_internal/integrations/llm_providers/anthropic.py [hunk 11]
```diff
@@ -117,38 +284,59 @@ def record_chunk(self, chunk: anthropic.types.MessageStreamEvent) -> None:
...
+            result['response_data'] = self.get_response_data()
+        if 'latest' in versions and self._content:
+            combined = ''.join(self._content)
+            result[OUTPUT_MESSAGES] = [
+                {
+                    'role': 'assistant',
+                    'parts': [TextPart(type='text', content=combined)],
+                }
+            ]
+        return result
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/anthropic.py:300]: Acknowledged — this is a pre-existing limitation inherited from the v1 streaming path, which also only captures text. Streaming tool-use output messages can be addressed in a follow-up.

File: logfire/_internal/integrations/llm_providers/openai.py [hunk 12]
```diff
@@ -117,24 +154,32 @@ def get_endpoint_config(options: FinalRequestOptions) -> EndpointConfig:
...
-            'events': inputs_to_events(
-                json_data.get('input'),
-                json_data.get('instructions'),
-            ),
             PROVIDER_NAME: 'openai',
             OPERATION_NAME: 'chat',
             REQUEST_MODEL: json_data.get('model'),
         }
+        if 1 in versions:
+            span_data['events'] = inputs_to_events(json_data.get('input'), json_data.get('instructions'))
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/llm_providers/openai.py]: Already fixed — the code at line 549 now uses `(span_data.get('events') or [])` instead of `span_data.get('events', [])`, which handles the case where `inputs_to_events` returns `None` (the `or []` coerces `None` to `[]`).

File: logfire/_internal/main.py [hunk 13]
```diff
@@ -1179,6 +1180,7 @@ def instrument_openai(
         | None = None,
         *,
         suppress_other_instrumentation: bool = True,
+        version: SemconvVersion | Sequence[SemconvVersion] = 1,
     ) -> AbstractContextManager[None]:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/main.py:1184]: No change needed — the logfire-api stubs already use `*args, **kwargs` signatures, so they accept any parameters including `version` without raising TypeError.

**Comment by dmontagu** [comment]
Closing in favor of #1705 because I'm having a hard time dealing with the fact that this PR is from a different fork and main is being updated in parallel, having it in the same repo I think will make things simpler.
